torchvision/csrc/io/image/cpu/decode_png.cpp

```diff
@@ -25,27 +25,41 @@ torch::Tensor decode_png(const torch::Tensor& data, ImageReadMode mode) {
   auto info_ptr = png_create_info_struct(png_ptr);
   if (!info_ptr) {
     png_destroy_read_struct(&png_ptr, nullptr, nullptr);
-    // Seems redundant with the if statement. done here to avoid leaking memory.
+    // Seems redundant with the if statement. Done here to avoid leaking memory.
     TORCH_CHECK(info_ptr, "libpng info structure allocation failed!")
   }
 
-  auto datap = data.accessor<unsigned char, 1>().data();
+  auto deleter = [&png_ptr, &info_ptr](png_structp __unused_ptr) {
+    png_destroy_read_struct(&png_ptr, &info_ptr, nullptr);
+  };
+  std::unique_ptr<png_struct, decltype(deleter)> png_ptr_RAII(png_ptr, deleter);
+
+  auto accessor = data.accessor<unsigned char, 1>();
+  auto datap = accessor.data();
+  auto data_size = accessor.size(0);
```

Comment:
nit: this can probably be simplified as `uint8_t datap = data.data_ptr<uint8_t>();`, but this was already like that before, so let's not change this here